src/binary_basic.ml

```diff
@@ -0,0 +1,471 @@
+open Types
+
+module Input = struct
+  type t =
+    { bytes : bytes
+    ; pt : int
+    ; size : int
+    }
+
+  let _pp input =
+    print_int input.pt;
+    print_string " ";
+    print_int input.size;
+    print_newline ()
+
+  let is_empty input = input.size = 0
+
+  let from_bytes bytes =
+    let size = Bytes.length bytes in
+    { bytes; pt = 0; size }
+
+  let sub ~pos ~len input =
+    if pos <= input.size && len <= input.size - pos then
+      { input with pt = input.pt + pos; size = len }
+    else failwith "Input.sub error"
+
+  let sub_suffix pos input = sub ~pos ~len:(input.size - pos) input
+
+  let sub_prefix len input = sub ~pos:0 ~len input
+
+  let get n input =
+    if n < input.size then Some (Bytes.get input.bytes (input.pt + n)) else None
+
+  let get0 = get 0
+end
+
+let read_byte input =
+  match Input.get0 input with
+  | None -> failwith "read_byte error"
```

Comment:
Could you start using the error monad ? Returning `Error "read_byte error"` will be enough to start.